planning/planning_scene_monitor/src/planning_scene_monitor.cpp

```diff
@@ -818,35 +821,70 @@ void planning_scene_monitor::PlanningSceneMonitor::currentWorldObjectUpdateCallb
     }
 }
 
-void planning_scene_monitor::PlanningSceneMonitor::syncSceneUpdates(const ros::Time &t)
+bool planning_scene_monitor::PlanningSceneMonitor::waitForCurrentRobotState(const ros::Time &t, double wait_time)
 {
   if (t.isZero())
-    return;
+    return false;
+  ros::WallTime start = ros::WallTime::now();
+  ros::WallDuration timeout(wait_time);
+
+  ROS_DEBUG_NAMED("PSM", "sync robot state to: %.3fs", fmod(t.toSec(), 10.));
+
+  if (current_state_monitor_)
+  {
+    // Wait for next robot update in state monitor. Those updates are only passed to PSM when robot actually moved!
+    enforce_next_state_update_ = true; // enforce potential updates to be directly applied
+    bool success = current_state_monitor_->waitForCurrentState(t, wait_time);
+    enforce_next_state_update_ = false; // enforce potential updates to be directly applied
+
+    /* If the robot doesn't move, we will never receive an update from CSM in planning scene.
+       As we ensured that an update, if it is triggered by CSM, is directly passed to the scene,
+       we can early return true here (if we successfully received a CSM update). Otherwise return false. */
+    if (success)
+      return true;
+
+    ROS_WARN("Failed to fetch current robot state.");
+    return false;
+  }
 
-  // Robot state updates in the scene are only triggered by the state monitor on changes of the state.
-  // Hence, last_state_update_time_ might be much older than current_state_monitor_ (when robot didn't moved for a while).
+  // Eventually there is no state monitor. In this case state updates are received as part of scene updates only.
```

Comment:
"Eventually" doesn't make sense here (false friend for Germans), "There is no state monitor" is enough.